nflow-tests/src/main/java/io/nflow/tests/demo/workflow/BulkWorkflow.java

```diff
@@ -0,0 +1,137 @@
+
+  public static final String BULK_WORKFLOW_TYPE = "bulk";
+
+  private final static EnumSet<WorkflowInstanceStatus> RUNNING_STATES = complementOf(EnumSet.of(finished, created));
+
+  private static final Logger logger = getLogger(BulkWorkflow.class);
+
+  @Inject
+  WorkflowInstanceService instanceService;
+
+  public enum State implements WorkflowState {
+    splitWork(start), waitForChildrenToFinish(normal), done(end), error(manual);
+
+    private WorkflowStateType type;
+
+    State(WorkflowStateType type) {
+      this.type = type;
+    }
+
+    @Override
+    public WorkflowStateType getType() {
+      return type;
+    }
+
+    @Override
+    public String getDescription() {
+      return name();
+    }
+  }
+
+  protected BulkWorkflow(String type) {
+    super(type, splitWork, error);
+    setDescription("Executes child workflows in bulk but gracefully without effecting non-bulk tasks.");
+    permit(splitWork, waitForChildrenToFinish);
+    permit(waitForChildrenToFinish, done);
+  }
+
+  public BulkWorkflow() {
+    this(BULK_WORKFLOW_TYPE);
+  }
+
+  public NextAction splitWork(StateExecution execution, @StateVar(value = "requestData", readOnly = true) JsonNode data) {
+    boolean childrenFound = splitWorkImpl(execution, data);
+    if (childrenFound) {
+      return moveToState(waitForChildrenToFinish, "Running");
+    }
+    return retryAfter(now().plusMinutes(1), "Waiting for child workflows");
+  }
+
+  protected boolean splitWorkImpl(StateExecution execution, @SuppressWarnings("unused") JsonNode data) {
+    if (execution.getAllChildWorkflows().isEmpty()) {
+      throw new RuntimeException(
+          "No child workflows found - either add them before starting the parent or implement splitWorkflowImpl");
+    }
+    return true;
+  }
+
+  public NextAction waitForChildrenToFinish(StateExecution execution,
+      @StateVar(value = "concurrency", readOnly = true) int concurrency) {
+    List<WorkflowInstance> childWorkflows = execution.getAllChildWorkflows();
+    long running = childWorkflows.stream().filter(this::isRunning).count();
+    long completed = childWorkflows.stream().filter(this::isFinished).count();
+    if (completed == childWorkflows.size()) {
+      return moveToState(done, "All children completed");
+    }
+    long toStart = min(max(1, concurrency) - running, childWorkflows.size() - completed);
+    if (toStart > 0) {
+      childWorkflows.stream().filter(this::isInInitialState).limit(toStart).forEach(this::wakeup);
+      logger.info("Started " + toStart + " child workflows");
+    }
+    long progress = completed * 100 / childWorkflows.size();
+    return retryAfter(now().plusMinutes(15), "Waiting for child workflows to complete - " + progress + "% done");
+  }
+
+  private void wakeup(WorkflowInstance instance) {
+    instanceService.wakeupWorkflowInstance(instance.id, emptyList());
+  }
+
+  private boolean isRunning(WorkflowInstance instance) {
```

Comment:
This should be `protected` so that extending class can override it - sometimes you want to start another workflow if the main flow goes to manual step for example.

Reply:
ok